akka-stream/src/main/scala/akka/stream/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import akka.Done
+import akka.actor.{ ActorRef, Terminated }
+import akka.event.Logging
+import akka.stream._
+import akka.stream.StreamRefs
+import akka.stream.impl.StreamRefsMaster
+import akka.stream.stage._
+import akka.util.{ OptionVal, PrettyDuration }
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.Try
+
+object SinkRef {
+  def source[T](): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(new SourceRef[T](OptionVal.None, canMaterializeSinkRef = true))
+
+  // TODO Implement using TCP
+  // steps:
+  // - lazily, but once bind a port
+  // - advertise to other side that they may send data into this port
+  // -- "passive mode" ftp ;-)
+  // def bulkTransferSource(port: Int = -1): Source[ByteString, SinkRef[ByteString]] = ???
+}
+
+/**
+ * The dual of SourceRef.
+ *
+ * This is the "handed out" side of a SinkRef. It powers a Source on the other side.
+ *
+ * Do not create this instance directly, but use `SinkRef` factories, to run/setup its targetRef.
+ *
+ * We do not materialize the refs back and forth, which is why the 2nd param.
+ */
+final class SinkRef[In] private[akka] (
+  private[akka] val initialPartnerRef:       OptionVal[ActorRef],
+  private[akka] val canMaterializeSourceRef: Boolean
+) extends akka.stream.javadsl.SinkRef[In] {
+  import akka.stream.StreamRefs._
+
+  val in = {
+    val inletName = s"${Logging.simpleName(getClass)}($initialRefName).in"
+    Inlet[In](inletName)
+  }
+  override def shape: SinkShape[In] = SinkShape.of(in)
+
+  private def initialRefName =
+    if (initialPartnerRef.isDefined) initialPartnerRef.get
+    else "<no-initial-ref>"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val promise = Promise[SourceRef[In]]
+    if (!canMaterializeSourceRef) promise.failure(StreamRefs.CyclicMaterializationAttemptException(
+      "SinkRef", otherStage = "SourceRef"))
```

Comment:
that's the current "dont keep materializing in circles all the time" protection // @jrudolph 
Pulling the same off statically gets painful since we need more types than just SourceRef SinkRef then, so it gets confusing for users I thought...

If we can find a way for this statically and not crazy that'd be nice

Reply:
Yep, agreed. I wonder if a cleaner separation between API and implementation would simplify getting there. I'll have a quick try if that's reasonable.

Reply:
Here's a simple potential solution: https://github.com/ktoso/akka/pull/33